compiler/vm/vmtypes.nim

```diff
@@ -0,0 +1,442 @@
+## This module contains the `TypeInfoCache` implementation together with procs
+## for type info lookup and creation
+
+import 
+  ast/[
+    ast_types,
+    ast,
+    reports,
+    types
+  ],
+  front/[
+    msgs
+  ],
+  vm/[
+    vmdef
+  ],
+
+  std/[
+    tables
+  ]
+
+
+import std/options as soptions
+
+from std/bitops import bitand, fastLog2
+
+
+template paddedSize(s, a: uint): untyped =
+  ## Round the size up so that `s % a == 0`
+  bitand(s + a - 1, not (a - 1))
+
+template paddedSize(info: VmType): untyped =
+  ## Round the size up so that `info.sizeInBytes % (1 shl info.alignment) == 0`
+  paddedSize(info.sizeInBytes, 1'u shl info.alignment)
+
+
+proc `==`*(a, b: VmType): bool =
+  # XXX: csources compiler doesn't support auto-generated `==` operator for case objects :(
+  if a.kind != b.kind or a.sizeInBytes != b.sizeInBytes or a.alignment != b.alignment:
+    return false
+
+  case a.kind
+  of akPtr, akRef:
+    a.targetType == b.targetType
+  of akSeq:
+    a.seqElemType == b.seqElemType and a.seqElemStride == b.seqElemStride
+  of akCallable:
+    a.retType == b.retType
+  of akObject:
+    a.objFields == b.objFields
+  of akArray:
+    # Ignore `elementStride`. If type is the same, so should the stride
+    a.elementCount == b.elementCount and a.elementType == b.elementType
+  else:
+    true
+
+
+func arrayType(len: int, elemTyp: PVmType): VmType =
+  # TODO: should two empty array types with differing element types 

```

Comment:
Only if being treated as opaque, otherwise no. I think what I'm saying is there is probably a rarely used array type that can generalize over arrays and it's a bit better than a void pointer.